Enable use of reference items in the DataBox

kidder opened the pull request:
## Proposed changes

This PR enables the use of reference items in the DataBox.

All compute tags that derive from `db::ComputeTag` except FromGlobalCache have been converted to mutating compute tags (their member function `function` returns void and takes their `return_type` by `gsl::not_null`).  

Non-mutating compute tags will be replaced with reference tags (i.e. derive from `db::ReferenceTag`)

This PR:
- Renames some type aliases in DataBox by using mutable and immutable instead of simple and compute as this in the important distinction between the tags in the type lists.  
- Replaces `db::is_compute_tag` with `db::is_immutable_item_tag` to allow reference items in the DataBox
- Converts FromGlobalCache from a compute tag to a reference tag
- Add a test for reference items
- Removes db::get_items and db::get_compute_items

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
- Compute tags must return by reference (i.e. have a void `function` taking their `return_type` by `gsl::not_null`).
- Anything referring to the old names of the type aliases in DataBox will need to use the new name.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

Note that the tags for subitems of compute items are currently placed in the `immutable_item_tags` list in DataBox despite not being marked with `db::ComputeTag`.  A subsequent PR will mark these tags with either `db::ComputeTag` or `db::ReferenceTag` as appropriate.  

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Rename compute_item_tags to immutable_item_creation_tags

Immutable item tags will include both compute item tags and reference item tags.
Creation tags are the tags passed to db::create or db::create_from when constructing
a DataBox.

kidder committed:
Rename compute_with_subitem_tags to immutable_item_tags

kidder committed:
Rename simple_item_tags to mutable_item_tags

kidder committed:
Rename simple_only_expanded_subitems_tags to mutable_subitem_tags

Also remove simple_subitems_tags as it was only used in one place.

kidder committed:
Update type aliases and type logic for using reference tags

Change some type aliases to refer to mutable and immutable item tags.
Change some uses of db::is_compute_tag to db::is_immutable_item_tag.

moxcodes commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1093,50 +1095,40 @@ SPECTRE_ALWAYS_INLINE constexpr auto create_from(Box&& box,
+  using immutable_item_tags_to_remove = expand_subitems<
+      tmpl::filter<remove_tags, db::is_immutable_item_tag<tmpl::_1>>>;
+  using mutable_item_tags_to_remove = expand_subitems<
+      tmpl::list_difference<remove_tags, immutable_item_tags_to_remove>>;
```
Can this be done with a filter with `tmpl::not_`? I haven't run any tests on this, but reasoning through the tmp operations, it seems that filtering with an easily compiled metafunction should be faster than `list_difference` if it gives the same result.

moxcodes commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -151,10 +151,10 @@ struct storage_type_impl {
-                                        : std::is_base_of_v<db::SimpleTag, Tag>
-                                              ? 1
-                                              : 0>::template f<TagList, Tag>;
+                : is_immutable_item_tag_v<Tag>
```
Should this change be in one of the renaming commits?

nilsvu commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -813,20 +813,21 @@ SPECTRE_ALWAYS_INLINE constexpr void
 
-  EXPAND_PACK_LEFT_TO_RIGHT(helper(Subtags{}, is_compute_tag<ParentTag>{}));
+  EXPAND_PACK_LEFT_TO_RIGHT(
+      helper(Subtags{}, is_immutable_item_tag<ParentTag>{}));
```
(optional) I think the overloaded lambda could be quite easily converted to an `if constexpr`.

kidder replied:
yes, that is coming in a forthcoming PR that changes other things about the function as well

nilsvu commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1019,15 +1020,17 @@ struct compute_dbox_type<tmpl::list<ItemsPack...>, ComputeItemsList> {
  */
 template <class TagList>
-using get_compute_items = tmpl::filter<TagList, db::is_compute_tag<tmpl::_1>>;
+using get_compute_items =
```
Should this metafunction and the one below also be renamed? In particular, in the docs for the function below I don't understand what "all Tags that are items" means.

kidder replied:
yes, I was planning to do that and forgot

moxcodes commented on tests/Unit/DataStructures/DataBox/Test_DataBox.cpp:
```diff
@@ -2674,6 +2674,33 @@ struct FromTaggedTuple : Tag, db::ReferenceTag {
 /// [databox_reference_tag_example]
 }  // namespace test_databox_tags
+
+void test_reference_item() noexcept {
```
This is a useful test, but also seems a little confusing in a commit about the `FromGlobalCache` tag.

moxcodes reviewed (commented):
I'm happy with you squashing any fixups directly.

nilsvu reviewed (commented):
Could you make the upgrade instructions a bit clearer? I.e., is there anything we should be aware of when rebasing our code on this PR once it's merged?

Are reference tags actually done and functional once this PR is merged? If so I suggest you reflect that in the PR title and in the upgrade instructions.

kidder committed:
Use tmpl::filter instead of list_difference in type alias

This should compile faster

kidder committed:
Remove db::get_items and db::get_compute_items

They are no longer needed as the list of simple_tags is now available in actions.

kidder committed:
Convert Parallel::Tags::FromGlobalCache to a reference tag

This is the only non-mutating compute item in the code.
Also enable handling reference tags in DataBox.

kidder committed:
Add test of reference item

kidder commented:
@nilsdeppe @nilsleiffischer @moxcodes 
- rebased and squashed
- added a new commit removing `db::get_items` and `db::get_compute_items` which are no longer needed
- updated the PR name and made upgrade instructions more explicit

nilsvu commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -254,7 +254,7 @@ template <typename Tag, typename... Tags, typename... TagTypes>
                                        tmpl::list<TagTypes...>> {
   static_assert(
-      is_function_callable_v<Tag, TagTypes...>,
+      is_get_callable_v<Tag, TagTypes...>,
```
I still don't quite understand the implications of this PR, and I think most of my confusion comes from symbol names that may just not be updated yet:

- We have "simple/compute/reference/base tags" now, where compute/reference tags are "immutable". Are simple and base tags treated separately from the "mutable/immutable" concept?
- Code that refers to either compute or reference tags is still named "compute" as far as I can tell, which is probably my main source of confusion. Should the struct I'm commenting on here and the one below carry "reference" in their name instead, and also say "reference item" in the error message?
- If I understand correctly, the change here and the one in DataBox.hpp:459(new) mean that compute items _must_ return by ref now. That seems the important thing to add to the upgrade instructions. (The renamed type aliases aren't used much anyway, so not sure if you even need to add them to the upgrade instructions.)
- The code around DataBox.hpp:459 has "mutable", "immutable" and "compute" in names, though if my understanding is correct it should only distinguish "compute" and "reference" (both of which are "immutable" by the definition of tags, though of course I understand that compute tags actually mutate storage in the DataBox so the choice of name might be confusing or should at least be commented on where it's relevant).

kidder replied:
There are two longer PRs coming next that should address all of these.  The first rewrites the implementation and the second updates the overall documentation.
-  Mutable vs immutable distinguish those items that can or cannot be mutated by calling `db::mutate`.  Compute and reference items are immutable; all others (which at the moment are simple items) are mutable.  
- All compute (reference) tags must derive from `db::ComputeTag` (`db::ReferenceTag`) and a simple tag.
- All simple tags must derive from `db::SimpleTag`.  A simple tag may also (in rare cases where type-erasure is needed) derive from a base tag (which itself must derive from `db::BaseTag`).
- The PR rewriting the implementation will also clean up the internal names, addressing your 2nd and 4th bullets.  
- Yes, your third bullet is true, I've modified the upgrade instructions

nilsvu replied:
Ok I'll wait for those next PRs then :)

kidder commented:
@nilsdeppe this needed to be rebased
